### PR TITLE
implement megavault shares locking

### PR DESCRIPTION
**Pull request by tqin7**
### Changelist
add mechanism to lock megavault shares
- will add shares unlocking in the next PR

### Test Plan
unit tests

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
	- Introduced properties for tracking locked shares and owner share unlocks in the vault's genesis state.
	- Added interfaces and methods for managing locked shares and their unlocking details, enhancing share management capabilities.
	- Implemented functions for retrieving and setting locked shares and unlocks, improving data handling.
	- Enhanced error handling for invalid locked shares and owner share unlocks to improve user feedback and maintain data integrity.
	- Introduced validation methods for owner share unlocks to ensure accuracy in data representation.

- **Bug Fixes**
	- Updated initialization and export processes to accurately include locked shares and their unlocks in the vault's state.

- **Tests**
	- Expanded test coverage for locked shares and owner share unlocks functionality, including validation, retrieval, and locking scenarios.

### CODE REVIEW

File: proto/dydxprotocol/vault/genesis.proto [hunk 1]
```diff
@@ -18,6 +18,8 @@ message GenesisState {
   repeated Vault vaults = 3 [ (gogoproto.nullable) = false ];
   // The default quoting parameters for all vaults.
   QuotingParams default_quoting_params = 4 [ (gogoproto.nullable) = false ];
+  // All locked shares.
```
Comment (vincentwschau) [inline proto/dydxprotocol/vault/genesis.proto:21]: nit: Add a note to `total_shares` about whether it includes or excludes locked shares.

File: protocol/x/vault/keeper/shares_test.go [hunk 2]
```diff
@@ -173,3 +173,192 @@ func TestGetAllOwnerShares(t *testing.T) {
...
+			expectedErr:          "invalid parameters",
+		},
+		"Error - negative shares to lock": {
+			existingLockedShares: nil,
+			ownerAddress:         constants.AliceAccAddress.String(),
+			sharesToLock:         big.NewInt(-1),
+			lockUntilBlock:       2,
+			expectedErr:          "invalid parameters",
+		},
+		"Error - lock until height same as current block height": {
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/shares_test.go:319]: nit: Also test for locking for a prior height.

File: proto/dydxprotocol/vault/share.proto [hunk 3]
```diff
@@ -21,3 +21,22 @@ message OwnerShare {
   string owner = 1 [ (cosmos_proto.scalar) = "cosmos.AddressString" ];
   NumShares shares = 2 [ (gogoproto.nullable) = false ];
 }
+
+// LockedShares stores for an owner their total number of locked shares
+// and a schedule of share unlockings.
+message LockedShares {
+  // Address of the owner of below shares.
+  string owner_address = 1;
+  // Total number of locked shares.
```
Comment (vincentwschau) [inline proto/dydxprotocol/vault/share.proto:30]: nit: Thought through this a bit more, I think we can remove this property, and have this be a function on the `LockedShares` type, that iterates through `unlock_details` and returns the total.
Once you have a pointer to the type, all the unlock details are unmarshalled anyways, so I don't think we save anything on serde by having this aggregated value in `LockedShares`, and the remaining iteration / arithmetic will be fast.

File: protocol/x/vault/keeper/shares.go [hunk 4]
```diff
@@ -90,3 +92,104 @@ func (k Keeper) GetAllOwnerShares(ctx sdk.Context) []types.OwnerShare {
...
+			TotalLockedShares: sharesToLock,
+			UnlockDetails: []types.UnlockDetail{
+				{
+					Shares:            sharesToLock,
+					UnlockBlockHeight: tilBlock,
+				},
+			},
+		}
+	} else {
+		// Increment total unlocked shares.
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/shares.go:175]: nit: `// Increment total locked shares `

File: protocol/x/vault/keeper/shares.go [hunk 5]
```diff
@@ -90,3 +92,104 @@ func (k Keeper) GetAllOwnerShares(ctx sdk.Context) []types.OwnerShare {
...
+	defer lockedSharesIterator.Close()
+	for ; lockedSharesIterator.Valid(); lockedSharesIterator.Next() {
+		var lockedShares types.LockedShares
+		k.cdc.MustUnmarshal(lockedSharesIterator.Value(), &lockedShares)
+		allLockedShares = append(allLockedShares, lockedShares)
+	}
+	return allLockedShares
+}
+
+// LockShares locks `sharesToLock` for `ownerAddress` until height `tilBlock`.
```
Comment (vincentwschau) [inline protocol/x/vault/keeper/shares.go:143]: nit: Should there be a TODO to compare the `sharesToLock` to the total number of shares owned by the `ownerAddress`?

File: proto/dydxprotocol/vault/share.proto (hunk 3 above)
Comment (tqin7) [reply proto/dydxprotocol/vault/share.proto:30]: Agreed that it's an extra thing to keep in mind to update both and it's simpler to iterate through to get total locked shares

**Review by vincentwschau — APPROVED** [review]

**Pull request merged**